modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java

```diff
@@ -0,0 +1,358 @@
+            LongMetric finished = (LongMetric)mreg.findMetric(FINISHED);
+            LongMetric totalExecutionTime = (LongMetric)mreg.findMetric(EXECUTION_TIME);
+            LongMetric totalWaitingTime = (LongMetric)mreg.findMetric(WAITING_TIME);
+
+            assertNotNull(started);
+            assertNotNull(active);
+            assertNotNull(waiting);
+            assertNotNull(canceled);
+            assertNotNull(rejected);
+            assertNotNull(finished);
+            assertNotNull(totalExecutionTime);
+            assertNotNull(totalWaitingTime);
+
+            assertEquals(0, started.value());
+            assertEquals(0, active.value());
+            assertEquals(0, waiting.value());
+            assertEquals(0, canceled.value());
+            assertEquals(0, rejected.value());
+            assertEquals(0, finished.value());
+            assertEquals(0, totalExecutionTime.value());
+            assertEquals(0, totalWaitingTime.value());
+
+            SimplestTask task1 = new SimplestTask();
+            SimplestTask task2 = new SimplestTask();
+            SimplestTask task3 = new SimplestTask();
+
+            task1.block = true;
+            task2.block = true;
+            task3.block = true;
+
+            //Task will become "waiting", because of CollisionSpi implementation.
+            ComputeTaskFuture<?> fut1 = g.compute().executeAsync(task1, 1);
+            ComputeTaskFuture<?> fut2 = g.compute().executeAsync(task2, 1);
+            ComputeTaskFuture<?> fut3 = g.compute().executeAsync(task3, 1);
+
+            assertEquals(0, started.value());
+            assertEquals(0, active.value());
+            assertEquals(3, waiting.value());
+            assertEquals(0, canceled.value());
+            assertEquals(0, rejected.value());
+            assertEquals(0, finished.value());
+
+            //Activating 2 of 3 jobs. Rejecting 1 of them.
+            Iterator<CollisionJobContext> iter = collisioinSpi.jobs.values().iterator();
+
+            iter.next().cancel();
+
+            assertEquals(1, rejected.value());
+
+            Thread.sleep(100); //Sleeping to make sure totalWaitingTime will become more the zero.
+
+            iter.next().activate();
+            iter.next().activate();
+
+            boolean res = waitForCondition(() -> active.value() > 0, TIMEOUT);
+
+            assertTrue(res);
+
+            Thread.sleep(100); //Sleeping to make sure totalExecutionTime will become more the zero.
+
+            latch.countDown();
+
+            res = waitForCondition(() -> fut1.isDone() && fut2.isDone() && fut3.isDone(), TIMEOUT);
+
+            assertTrue(res);
+
+            res = waitForCondition(() -> finished.value() == 3, TIMEOUT);
+
+            assertTrue(res);
+
+            assertTrue("Execution time should be greater then zero.", totalExecutionTime.value() > 0);
+            assertTrue("Waiting time should be greater then zero.", totalWaitingTime.value() > 0);
```

Comment:
It's better to check `totalWaitingTime` earlier, when `totalExecutionTime` is zero

Reply:
Check moved to the line 143